src/target/riscv/riscv-013.c

```diff
@@ -2811,6 +2818,320 @@ void riscv013_fill_dmi_nop_u64(struct target *target, char *buf)
+		target_addr_t legal_address, uint32_t num_words,
+		target_addr_t illegal_address, bool run_sbbusyerror_test)
+{
+	LOG_INFO("Testing System Bus Access as defined by RISC-V Debug Spec v0.13");
+
+	uint32_t rd_val;
+	uint32_t sbcs_orig;
+	dmi_read(target, &sbcs_orig, DMI_SBCS);
+
+	uint32_t sbcs = sbcs_orig;
+	bool test_passed;
+
+	int max_sbaccess = get_max_sbaccess(target);
+
+	if (max_sbaccess == -1) {
+		LOG_ERROR("System Bus Access not supported in this config.");
+		return ERROR_FAIL;
+	}
+
+	if (get_field(sbcs, DMI_SBCS_SBVERSION) != 1) {
+		LOG_ERROR("System Bus Access unsupported SBVERSION (0). Only version 1 is supported.");
+		return ERROR_FAIL;
+	}
+
+	uint32_t num_sbdata_regs = get_num_sbdata_regs(target);
+
+	uint32_t rd_buf[num_sbdata_regs];
+
+	/* Test 1: Simple write/read test */
+	test_passed = true;
+	sbcs = set_field(sbcs_orig, DMI_SBCS_SBAUTOINCREMENT, 0);
+	dmi_write(target, DMI_SBCS, sbcs);
+
+	uint32_t test_patterns[4] = {0xdeadbeef, 0xfeedbabe, 0x12345678, 0x08675309};
+	for (uint32_t sbaccess = 0; sbaccess <= (uint32_t)max_sbaccess; sbaccess++) {
+		sbcs = set_field(sbcs, DMI_SBCS_SBACCESS, sbaccess);
+		dmi_write(target, DMI_SBCS, sbcs);
+
+		uint32_t compare_mask = (sbaccess == 0) ? 0xff : (sbaccess == 1) ? 0xffff : 0xffffffff;
+
+		for (uint32_t i = 0; i < num_words; i++) {
+			uint32_t addr = legal_address + (i << sbaccess);
+			uint32_t wr_data[num_sbdata_regs];
+			for (uint32_t j = 0; j < num_sbdata_regs; j++)
+				wr_data[j] = test_patterns[j] + i;
+			write_memory_sba_simple(target, addr, wr_data, num_sbdata_regs, sbcs);
+		}
+
+		for (uint32_t i = 0; i < num_words; i++) {
+			uint32_t addr = legal_address + (i << sbaccess);
+			read_memory_sba_simple(target, addr, rd_buf, num_sbdata_regs, sbcs);
+			for (uint32_t j = 0; j < num_sbdata_regs; j++) {
+				if (((test_patterns[j]+i)&compare_mask) != (rd_buf[j]&compare_mask)) {
+					LOG_ERROR("System Bus Access Test 1: Error reading non-autoincremented address %x,"
+							"expected val = %x, read val = %x", addr, test_patterns[j]+i, rd_buf[j]);
+					test_passed = false;
+				}
+			}
+		}
+	}
+	if (test_passed)
+		LOG_INFO("System Bus Access Test 1: Simple write/read test PASSED.");
+
+	/* Test 2: Address autoincrement test */
+	target_addr_t curr_addr;
+	target_addr_t prev_addr;
+	test_passed = true;
+	sbcs = set_field(sbcs_orig, DMI_SBCS_SBAUTOINCREMENT, 1);
+	dmi_write(target, DMI_SBCS, sbcs);
+
+	for (uint32_t sbaccess = 0; sbaccess <= (uint32_t)max_sbaccess; sbaccess++) {
+		sbcs = set_field(sbcs, DMI_SBCS_SBACCESS, sbaccess);
+		dmi_write(target, DMI_SBCS, sbcs);
+
+		dmi_write(target, DMI_SBADDRESS0, legal_address);
+		read_sbcs_nonbusy(target, &sbcs);
+		curr_addr = legal_address;
+		for (uint32_t i = 0; i < num_words; i++) {
+			prev_addr = curr_addr;
+			read_sbcs_nonbusy(target, &sbcs);
+			curr_addr = sb_read_address(target);
+			if ((curr_addr - prev_addr != (uint32_t)(1 << sbaccess)) && i != 0) {
```

Comment:
I'm a fan of parens around this sort of thing (the curr_add - prev_addr) and the two sides of the && but not sure what the OpenOCD style guide is